secio/src/dh_compat/openssl_impl.rs

```diff
@@ -0,0 +1,177 @@
+            KeyAgreement::X25519 => &X25519,
+            #[cfg(not(ossl110))]
+            _ => panic!("ECDH {:?} does not supported by openssl yet", src),
+        }
+    }
+}
+
+pub struct EphemeralPrivateKey {
+    evp_key: PKey<Private>,
+    al: &'static Algorithm,
+}
+
+fn ec_generate(nid: Nid) -> Result<(PKey<Private>, Vec<u8>), SecioError> {
+    let group = EcGroup::from_curve_name(nid)?;
+
+    let mut ctx = BigNumContext::new()?;
+
+    let ec_key = EcKey::generate(&group)?;
+
+    let pubkey =
+        ec_key
+            .public_key()
+            .to_bytes(&group, PointConversionForm::UNCOMPRESSED, &mut ctx)?;
+
+    let evp_key = PKey::from_ec_key(ec_key)?;
+    Ok((evp_key, pubkey))
+}
+
+fn ec_from_pubkey(nid: Nid, other_public_key: &[u8]) -> Result<PKey<Public>, SecioError> {
+    let group = EcGroup::from_curve_name(nid)?;
+    let mut ctx = BigNumContext::new()?;
+
+    let point = EcPoint::from_bytes(&group, other_public_key, &mut ctx)?;
+
+    let ec_key = EcKey::from_public_key(&group, &point)?;
+    PKey::from_ec_key(ec_key).map_err(Into::into)
+}
+
+fn p256_generate() -> Result<(PKey<Private>, Vec<u8>), SecioError> {
+    ec_generate(Nid::X9_62_PRIME256V1)
+}
+
+fn p256_from_pubkey(other_public_key: &[u8]) -> Result<PKey<Public>, SecioError> {
+    ec_from_pubkey(Nid::X9_62_PRIME256V1, other_public_key)
+}
+
+fn p384_generate() -> Result<(PKey<Private>, Vec<u8>), SecioError> {
+    ec_generate(Nid::SECP384R1)
+}
+
+fn p384_from_pubkey(other_public_key: &[u8]) -> Result<PKey<Public>, SecioError> {
+    ec_from_pubkey(Nid::SECP384R1, other_public_key)
+}
+
+#[cfg(ossl110)]
+fn x25519_generate() -> Result<(PKey<Private>, Vec<u8>), SecioError> {
+    let evp_key = PKey::generate_x25519()?;
+
+    let pubkey = evp_key.as_ref().raw_public_key()?;
+
+    Ok((evp_key, pubkey))
+}
+
+#[cfg(ossl110)]
+fn x25519_from_pubkey(pubkey: &[u8]) -> Result<PKey<Public>, SecioError> {
+    PKey::public_key_from_raw_bytes(pubkey, Id::X25519).map_err(Into::into)
+}
+
+pub fn generate_agreement(
+    algorithm: KeyAgreement,
+) -> Result<(EphemeralPrivateKey, Vec<u8>), SecioError> {
+    let al: &'static Algorithm = algorithm.into();
+
+    let (evp_key, ecdh_public_key) = (al.pairs_generate)()?;
+
+    Ok((EphemeralPrivateKey { evp_key, al }, ecdh_public_key))
+}
+
+pub fn agree(
+    _algorithm: KeyAgreement,
+    my_private_key: EphemeralPrivateKey,
+    other_public_key: &[u8],
+) -> Result<Vec<u8>, SecioError> {
+    if other_public_key.len() != my_private_key.al.pubkey_len {
+        return Err(SecioError::SecretGenerationFailed);
+    }
+
+    let peer_evp_key = (my_private_key.al.from_pubkey)(other_public_key)?;
+
+    let mut deriver = Deriver::new(&my_private_key.evp_key)?;
+
+    deriver.set_peer(&peer_evp_key)?;
+
+    let len = deriver.len()?;
+    let mut secret = Vec::with_capacity(len);
+    // Safety: capacity == len
+    unsafe {
+        secret.set_len(len);
+    }
+
+    deriver.derive(&mut secret)?;
+
+    Ok(secret)
```

Comment:
```suggestion
Ok(deriver.derive_to_vec()?)
```

Reply:
It's hard to weigh the usage issues here, from the code, `derive_to_vec` has one more step to assign a zero value than the current use. But when I benchmark the two separately, the difference is minimal, and I'm not quite sure if I need to modify

Reply:
it's UB to read data in an uninitialized vec, I prefer a initialized version here.

Reply:
Although there is no read for uninitialized vec here, and I have no particular preference for it